### ISSUE TITLE
Ctrl+F support, finding text on a page (even when scrolled off screen)

### ISSUE THREAD

**Issue by mariamhas** [flutter/flutter#65504]
Current status: https://github.com/flutter/flutter/issues/65504#issuecomment-1475059683

---

cannot browser search (CTRL +F) to find text on a page

**Comment by yjbanov** [comment]
Some ideas from @sidrao2006 

> ## Use case
> 
> Add support for browser search
> 
> ## Proposal
> 
> 1. Let browser handle the search
> 
> 2. Provide a widget specifically to support static text features 
> 
> 3. Let html handle semantics and rendering for such elements 
> 
> 4. Use the Intersection Observer API (https://developer.mozilla.org/en-US/docs/Web/API/Intersection_Observer_API) to detect searches. Also add style change listeners to the text, such that it won't be triggered when the framework changes the style of the text but will be triggered when the browser changes the text style (since the browser changes the text style to highlight the searched words)
> 
> 5. Once the framework knows that the browser is navigating to a certain piece of text, render the relevant elements/widgets which are supposed to be in the surroundings of the text
> 
> 6. The static text rendering widget should be low level and abstracted under the Text widget such that static text mide will be enabled automatically for texts that do not have dynamic dependencies.

**Comment by yjbanov** [comment]
Just some brainstorming:

Browser native search may not be the best fit for Flutter due to lazy rendering, and because it's unclear how the same functionality can be achieved on mobile and desktop. A custom search might be a better fit because it would work on all platforms. We should make sure that developers can easily implement text search in their apps. Perhaps the framework could assist somehow, I'm not sure. A quick survey of non-Flutter web apps shows that many developers prefer custom search bars, e.g. Twitter, Gmail, Discord, Docs, Calendar, Slack, Chat, Maps, Facebook. Many actively redirect the Ctrl + F shortcut to their custom search bars. In some text heavy cases both can be used (and are useful): Github, Amazon.com.

Mobile users frequently don't have keyboards, so there's no way to intercept Ctrl + F. If there's no app-specific search bar, they would invoke "Find in page..." (also available on desktop browsers, if you know where to find it in the menus). I imagine, however, that users would prefer the custom search bar, especially if it's sufficiently prominent on the UI.

**Comment by nt4f04uNd** [comment]
is it possible to propose changes to the html to add hooks for this? lack of native browser "Find in page..." on mobiles cannot be really solved by embedding some custom search. users don't know what we used to build our website and won't understand why this feature is suprisingly not working

**Comment by vidklopcic** [comment]
This is also a frequent customer complaint in our use-cases. As a user myself I hate websites that override CTRL+F functionality (which is considered a bad practise). It's really convenient and it's also a quick way to navigate around websites.

However, I believe that with canvaskit rendering it is impossible to implement native text search, as the text isn't (and won't be) rendered by the browser natively. Maybe the next best thing would be to have an easy to use implementation of a custom search bar which would:
- imitate standard search behaviour (highlight any matching text that was rendered - if we have lazy loaded lists etc.. it is perfectly fine and expected to have no matches
- after exiting the search, focus should move to the last highlighted widget / text (if there is any focusable parent)

For mobile devices there is one way to at least show a button for the in-app / emulated search bar when user enters a query into the default "find on page" bar:

https://user-images.githubusercontent.com/8578790/120169535-5b26f380-c200-11eb-8624-9e03642ecae1.mov

If we add all letters of the alphabet to a hidden section at the bottom, the browser will scroll to it when any letter is typed into the search box. It is a bit hackish, but I don't see any other way to achieve this.

**Comment by vidklopcic** [comment]
It might be even possible to capture the native searches by observing scroll events

https://user-images.githubusercontent.com/8578790/120180225-cb874200-c20b-11eb-9449-44c01b1fdc4c.mov

https://gist.github.com/vidklopcic/766951868068d666fd3f0e916d9bf602

**Comment by fzyzcjy** [comment]
any updates...?

**Comment by guplem** [comment]
Will this be feature be added?

**Comment by masus04** [comment]
I've been trying to implement a custom search bar, but even this workaround currently seems to be impossible for focus reasons:
https://github.com/flutter/flutter/issues/111470

Does anyone know how to implement a custom search field that can be focused when `CTRL` + `F` is pressed anywhere on the page?

**Comment by archie-sh** [comment]
Hi, any plans for the flutter team to tackle this? We use flutter web in production and our users are asking. Thanks.

**Comment by rodrigo-itao** [comment]
I've changed the default renderer to HTML and now the search with CTRL + F is working. 

https://docs.flutter.dev/development/platform-integration/web/renderers

I know it's a workaround and not a perfect solution, but it's working and my clientes are happy with it.

**Comment by guplem** [comment]
I think that to fix this, they could go with two approaches: 
 A) Overriding the native Ctrl+F functionality of the browser (something that looks like a Flutter way of doing it, so it would work cross-platform)
 B) Actually exporting the text of the app to "HTML readable" code. This would probably help with [the SEO issue](https://github.com/flutter/flutter/issues/46789), but I really don't know how you could link that search results to a visual highlight (and scrolling to that location) in the rendered Flutter code (JS canvas).

**Comment by yjbanov** [comment]
Based on popular demand, I'm putting this back into triage queue for reprioritization. I'm also forwarding this to the framework team due to the following:

- Due to lazy rendering, the engine only sees a subset of the UI that's currently visible. For proper text search functionality we should have the ability to scroll to the desired content, but scrolling is a framework-level concept.
- There's no link from layers and pictures back to widgets, so even if we did search them it's not clear what to do next. We should provide some level of interaction with the found content such as highlighting it, selecting it, etc.
- Searching UI should not be limited to web. It should work on AOT desktop and mobile too.
- The search should follow some logical widget order, similar to cross-widget text selection. This information lives in the framework.

In AOT desktop and mobile, and on desktop web, this function can be implemented via a keyboard shortcut. Mobile web will be tricky though due to reasons outlined in https://github.com/flutter/flutter/issues/65504#issuecomment-790820972. Here's an idea though:

- The framework sends a list of text strings that can be searched with unique widget IDs to the engine.
- The engine creates a hidden scrollable area and puts the text strings in it, such that "Find in page" can find them.
- When the user finds a string using the browser's built-in search function, the browser scrolls to the string.
- The engine listens to the scroll offset and maps it back to the original text string and the corresponding widget ID. This mapping can be made precise if the strings are spaced apart more than the height of the scroll viewport. This way a scroll offset value uniquely identifies the string that's found. The engine then sends a message back to the framework so that the framework scrolls the desired widget into view.

A problem with this approach is when the search text spans more than one string, we can only find the first string.

As an MVP we could start with a pure framework implementation that uses a keyboard shortcut. Integration with browser search can attempted as a follow-up.

**Comment by Levi-Lesches** [comment]
Adding to this that building/running with `--web-renderer html` does make this issue a _little_ better -- it still won't find something that's been scrolled off-screen, but it can find and highlight text that is on-screen.

**Comment by omar-selo** [comment]
Even with `--web-renderer html` it can't seem to find special characters. So if you're search string has for instance ' ', ':', '_', etc. it won't find it. Even if it's on-screen.

**Comment by oberoivarun** [comment]
We've had issues with getting google ads approved for our website because the verifier could not find our given contact information on the website. He was doing ctrl+f for that info. We really need this, we are too far gone into development to re-build our website using another framework.

**Comment by oleksii-pimenov** [comment]
any updates on this guys, pretty critical feature

**Comment by gsusI** [comment]
Do they realize that this is the single biggest downside of using Flutter? If it were resolved, I'd use it for most projects. It's crazy 😕.

**Comment by fleytman** [comment]
Dear @yjbanov,

The ability to search within a page using browser commands, such as Ctrl+F, is a fundamental feature that operates seamlessly across most websites. This issue has been open for four years, gathered significant community support with over 100 likes, and has multiple duplicate threads.

Could you please provide an update on whether we can anticipate a resolution to this problem in the coming months?

Thank you for your attention to this matter.

**Comment by cawfeecoder** [comment]
Is there a chance if someone we're to step forward and financially sponsor the feature (e.g. $1000), we could get it in?

**Comment by awsms** [comment]
> Is there a chance if someone we're to step forward and financially sponsor the feature (e.g. $1000), we could get it in?

+1, I would definitely contribute to this financially.

**Comment by mrsrmn** [comment]
Any updates? It's a huge problem on a huge platform like Flutter, CTRL + F is fundamental for every website. @yjbanov

**Comment by matthew-carroll** [comment]
My team at the Flutter Bounty Hunters is always available if there are financial backers.

I should point out that the amount of labor needed to actually propose, build, and land this in Flutter would be well beyond $1,000. This is largely due to Flutter's review process - both for design docs and for PRs. I would estimate something like $5,000 => 50 hours.

I'll also offer to record and publish a video of the work along the way, including drafting the design doc.

PS - For those focused on web, this might not be possible to implement due to weird browser controls over IME capabilities.

**Comment by dsyrstad** [comment]
@matthew-carroll 
> PS - For those focused on web, this might not be possible to implement due to weird browser controls over IME capabilities.

This issue seems primarily focused on web, and it even has label:platform-web. It seems like its doable, but with quite a bit of work.

**Comment by matthew-carroll** [comment]
@dsyrstad - I think it depends on the specific requirements. There's definitely some level of support that could be brought to web. For example, we can definitely catch the shortcut for `ctrl+f` and `cmd+f`. From there, we can definitely query a content index that we assemble within Flutter.

I was referring more to any browser-level behaviors that might happen without any interaction with JS APIs. I'm not sure all the things browsers do for "Find", but think about browsers where you right-click on a text field or something like that. You get a "native" browser popup. That behavior isn't exposed to any JS APIs. It's done automatically by the browser, completely outside of the developer context. Things like that are simply outside the scope of what Flutter, or any other UI toolkit, can interact with. You either use those DOM elements or you're SOL.

**Comment by dsyrstad** [comment]
@matthew-carroll 
> I was referring more to any browser-level behaviors ... all the things browsers do for "Find", but think about browsers where you right-click on a text field or something like that...

Agreed. This issue seems to focus on Ctrl+F, so it'd be great to have something for that - the 80/20 rule.

**Comment by matthew-carroll** [comment]
In that case, happy to do it with financial backing. The shortcut is easy. We'll need a design doc to cover the intricacies of assembling the content/text index from the widget tree.

**Comment by KimGenius** [comment]
How to korean text??
i can't search korean text like that (renderer option is html)
![스크린샷 2024-07-17 16 09 47](https://github.com/user-attachments/assets/975af90a-4ec4-4df1-9932-e12fa32bb3ce)

**Comment by yjbanov** [comment]
Triage: acknowledged

**Comment by AlexanderLid** [comment]
The fact that this isn't clearly stated in the documentation makes it feel like Google is trying to quietly sweep it under the rug to avoid making Flutter Web look bad. Standard web functionality like Ctrl+F can even be an absolute dealbreaker for apps where this is essential. Clear documentation on such limitations is crucial for setting the right expectations. It would take literally just a sentence in the documentation to save developers hours of frustration and set clear expectations.

**Comment by jnorkus** [comment]
I don't know how Flutter triage works but this is absolutely a must have for a web framework to be viable. Unfortunately this one is what made our team vote against Flutter and go back to Angular.

**Comment by desmond206x** [comment]
This is insane. https://github.com/input-output-hk/catalyst-voices/issues/888#issuecomment-2385079269

**Comment by cristaloleg** [comment]
@omar-selo is this still valid? `flutter build web --web-renderer html` and search for special symbols/emojis works perfect.

**Comment by jnorkus** [comment]
> @omar-selo is this still valid? `flutter build web --web-renderer html` and search for special symbols/emojis works perfect.

Now try to add space to your search. Or search for something that is off screen and not currently visible.

**Comment by dsyrstad** [comment]
> @omar-selo is this still valid? `flutter build web --web-renderer html` and search for special symbols/emojis works perfect.

@cristaloleg: The html renderer is going to be removed: #145954,  and it doesn't work at all with canvaskit.

**Comment by cristaloleg** [comment]
WOW, this sucks even more, thanks for the update 😢

**Comment by Tananga** [comment]
Yeah... CTRL F ussue together with SEO is the biggest problems in Flutter Web

**Comment by davidbt1** [comment]
I imagine the importance of this feature is already clear. Hundreds of people have taken part in this issue.
However, the importance of this feature is doubly significant for those already in production who forgot to take it into account because they assumed it was a given.
Over the years, this technology hasn't ceased to impress many of us. Its development has been meteoric, and we shouldn't let an issue like this hold it back.

**Comment by Terranic** [comment]
I think, at least using the browser search the visible text only would be a helpful feature. 
Of course the full version must be a complete search function..maybe a specific parent widget would help to collect all text and its position to make it visible to the browser search, although not yet fully rendered?

**Comment by nrajput05** [comment]
Both CTRL + F and CTRL + P are such important features for web and both are not working :(

**Comment by justinmc** [comment]
I was wondering if it would be possible to implement this in such a way that unrendered text in ListView.builder would also be findable. It's exciting to think about making this work even better than it works on web... But after thinking about it, it might not be a good idea... Text below an infinite ListView.builder would never be findable, for example.

**Comment by dsyrstad** [comment]
> I was wondering if it would be possible to implement this in such a way that unrendered text in ListView.builder would also be findable. It's exciting to think about making this work even better than it works on web... But after thinking about it, it might not be a good idea... Text below an infinite ListView.builder would never be findable, for example.

@justinmc Text in lazy loaded tables and lists in regular web apps is not findable by the browser currently. I'd be all for having it work just as good as regular web apps, instead of the current state where it doesn't work at all.

**Comment by TJ-coding** [comment]
It's kind of funny how little attention this thread has despite the seriousness of the issue. It makes me question if flutter is actually dead. There is a thread open on Flock as well with also very little attention.

**Can someone recommend a flutter package** that people can use to implement an  ad hoc `ctr + f` on flutter web as I think it will take another 5 years to get this feature officially implemented. Or maybe never

**Comment by clotodex** [comment]
I dont understand what is blocking this. So I made a naive list and am looking to spark a discussion to move this forward (or be educated :D)

How to design the API? -> probably many approaches possible, e.g. a `SelectionArea` type approach. This way developers can choose the potential performance penalty they want to give to the app - wrapping it around the root node or not. But this is "just" a design decision.

Native being impossible? -> I get this together with SEO for websites, but most WebApps have no issue with a "search override" approach (see docs, figma, notion, ...), me included

The actual "what happens" part? -> we need a scroll, and a focus, I am also happy to have something that does not solve everything initially and improves gradually. Like just notifying the user that there IS a match, but outside the screen, and they need to manually navigate there.

The highlighting? -> Initially: overlay builder at coordinate (but likely wont perfectly overlay letters etc due to fonts - or might be complicated to align), then we can move to box decorations, maybe via something that is injected by the `SelectionArea` type approach, before ideally also adding an option in a widget to override the "match found" functionality, where the widget itself can choose how it wants to handle the visualization.

Right now I feel like a simple package could not provide anything properly usable (might be mistaken) - the amount of different parts of flutter one needs to touch as a package developer and MOST importantly the API standard that others can implement etc can imho only be pushed via a flutter core. I think a lot (like how to visualize, special behavior, advanced features, etc) can be implemented by the community once a core direction is set by flutter.

**Comment by forumics** [comment]
if you used selectablearea to create text then you used a mouse to highlight the text, flutter will create a dom element in the background so that text is findable but won't be selected visually. when scrolled the position is updated according to the place it exists on html.
if there is a function that we can call upon ctrl + f to create dom element, it would be easy to create the visual highlighting part of search.

**Comment by adar2378** [comment]
Do you have any updates on this? Is anyone working? This is such a basic feature :/

**Comment by tananga** [comment]
@adar2378 maybe not basic but really important. Without ctrl + f and better SEO flutter web will always be used just for admin panels and internal tools

**Comment by piyushpatel130689** [comment]
Is there any solution for this issue?

**Comment by anggaaryas** [comment]
Wait, why can we select text using the selection area but still not find it? Can we intercept the browser’s find feature and integrate it with the selection area?

**Comment by guplem** [comment]
This has actually been suggested earlier in the thread (see [@forumics](https://github.com/flutter/flutter/issues/65504#issuecomment-2910005454) and [@clotodex](https://github.com/flutter/flutter/issues/65504#issuecomment-2903442575) comments). Seems technically feasible, just waiting on prioritization from the Flutter team.

**Comment by raphael-bmec-co** [comment]
Reminder that upvotes are a contributing factor to prioritisation. So if you haven't already, go here and 👍 https://github.com/flutter/flutter/issues/65504#issue-697156659 .

**Comment by anggaaryas** [comment]
About Selection Area, I hope the Team also fix this [issue](https://github.com/flutter/flutter/issues/153004) . Look like it will affect this topic too

**Comment by Yusufihsangorgel** [comment]
This will not be the framework-level fix people are asking for here (the browser's own Ctrl+F still cannot reach a Flutter canvas), but if you need in-page find working today, I published a small package for it: find_in_page (https://pub.dev/packages/find_in_page). You wrap text in a FindableText, drop in a FindBar, and it does the match, highlight, and next/previous loop.

The part relevant to this thread is the "even when scrolled off screen" case in the title. A ListView.builder only builds the rows near the viewport, so a naive find misses everything else and the match count drifts as you scroll. The latest version adds a FindableListView that registers every row's text up front from the backing data, so a match on a row that was never built is still found and counted, and it scrolls to reveal it. Two honest caveats: it needs a fixed itemExtent for that (the same constraint ListView.builder(itemExtent:) already has), and it searches the text you hand it rather than intercepting the browser, so it is a stopgap, not the real fix. It has covered the "long list, find something off-screen" need in my own apps.

Happy to hear if it misses a case you have.
